terraform/modules/dandiset_bucket/main.tf

```diff
@@ -142,60 +137,52 @@ resource "aws_s3_bucket_policy" "dandiset_bucket_policy" {
 data "aws_iam_policy_document" "dandiset_bucket_policy" {
   version = "2008-10-17"
 
-  dynamic "statement" {
-    for_each = var.public ? [1] : []
-
-    content {
-      resources = [
-        "${aws_s3_bucket.dandiset_bucket.arn}",
-        "${aws_s3_bucket.dandiset_bucket.arn}/*",
-      ]
+  statement {
+    resources = [
+      "${aws_s3_bucket.dandiset_bucket.arn}",
+      "${aws_s3_bucket.dandiset_bucket.arn}/*",
+    ]
 
-      actions = [
-        "s3:Get*",
-        "s3:List*",
-      ]
+    actions = [
+      "s3:Get*",
+      "s3:List*",
+    ]
 
-      principals {
-        identifiers = ["*"]
-        type        = "*"
-      }
+    principals {
+      identifiers = ["*"]
+      type        = "*"
     }
   }
 
   # Disallow access to embargoed objects, unless using the heroku user arn, or
   # an extra, authorized embargo reader account.
-  dynamic "statement" {
-    for_each = var.public ? [1] : []
-
-    content {
-      effect = "Deny"
-      principals {
-        identifiers = ["*"]
-        type        = "*"
-      }
-      actions = ["s3:*"]
-      resources = [
-        "${aws_s3_bucket.dandiset_bucket.arn}/*",
-      ]
-      condition {
-        test     = "StringEquals"
-        variable = "s3:ExistingObjectTag/embargoed"
-        values   = ["true"]
-      }
-      condition {
-        test     = "ArnNotEquals"
-        variable = "aws:PrincipalArn"
-        values = flatten([
-          var.heroku_user.arn,
-          [for user in var.embargo_readers : user.arn],
-        ])
-      }
+  statement {
+    effect = "Deny"
+    principals {
+      identifiers = ["*"]
+      type        = "*"
+    }
+    actions = ["s3:*"]
+    resources = [
+      "${aws_s3_bucket.dandiset_bucket.arn}/*",
+    ]
+    condition {
+      test     = "StringEquals"
+      variable = "s3:ExistingObjectTag/embargoed"
+      values   = ["true"]
+    }
+    condition {
+      test     = "ArnNotEquals"
+      variable = "aws:PrincipalArn"
+      values = flatten([
+        var.heroku_user.arn,
+        [for user in var.embargo_readers : user.arn],
+      ])
     }
   }
 
   dynamic "statement" {
-    for_each = var.allow_cross_account_heroku_put_object ? [1] : []
+    for_each = data.aws_caller_identity.sponsored_account.account_id != data.aws_caller_identity.current.account_id ? [1] : []
```

Comment:
Why keep this conditional at all? Will it be a problem to allow the staging bucket to have this permission, even if it's unnecessary?